src/variable.c

```diff
@@ -12,224 +12,315 @@
 #include <mruby/variable.h>
 #include <mruby/presym.h>
 
-struct iv_elem {
-  mrb_sym key;
-  mrb_value val;
-};
-
-/* Instance variable table structure */
+/*
+ * Instance variable table structure
+ *
+ * Layout:
+ *                                           iv_tbl* --.
+ *                                                      \
+ *           +---------+---------+---+-------+-------+---o--------+--------+
+ *   Type    |mrb_value|mrb_value|...|mrb_sym|mrb_sym|...|uint16_t|uint16_t|
+ *           +---------+---------+---+-------+-------+---+--------+--------+
+ *   Content |  val 1  |  val 2  |...| sym 1 | sym 2 |...|  size  | capa_1 |
+ *           +---------+---------+---+-------+-------+---+--------+--------+
+ *
+ *   - Only `val` and only `sym` are contiguous to eliminate structure padding.
+ *   - `val` is placed at the beginning to align it at 8-byte boundary.
+ *   - `capa_1` means "capacity - 1".
+ */
 typedef struct iv_tbl {
-  size_t size;
-  size_t alloc;
-  struct iv_elem *table;
+  char dummy1;
+  char dummy2[];
```

Comment:
I am not sure why we should do this kind of tricky memory allocation. Naive `struct` definition like:

```c
struct iv_tbl {
  uint16_t size;
  unit16_t capa_1;
};
```

and locating `values` and `syms` **after** `iv_tbl` seem to work. So we can access `size` and `capa_1` directly without tricky typecasting. Is there any reason?

Reply:
There are two reasons:

* To align `val` to 8-byte boundary (we had [a similar problem](https://github.com/mruby/mruby/pull/1353#issuecomment-20719140) in khash). I don't think `val` can aligned to 8-byte boundary if `size` and` capa_1` are placed at the beginning.
* It is easy to calculate address if `sizeof(iv_tbl)` is 1.

Reply:
@shuujii,
If the structure is the one shown by @matz, then `val` will be placed on an 8-byte boundary.
If you are concerned about padding and want to keep this PR layout, you should be fine with a structure that defines only `size` and `capa_1`.

```c
struct iv_tbl {
  /* mrb_value val[]; */
  /* mrb_sym sym[]; */
  uint16_t size;
  uint16_capa_1;
};
```

Example of an actual change (diff):
https://github.com/shuujii/mruby/compare/reduce-memory-usage-of-instance-variable-table...dearblue:shuujii/reduce-memory-usage-of-instance-variable-table+

Reply:
@dearblue, thank you for your comment.

> If the structure is the one shown by @matz, then `val` will be placed on an 8-byte boundary.

Ah, sorry, I misread the layout. `size -> capa_1 -> syms -> vals` has a boundary problem, but `size -> capa_1 -> vals -> syms` has no problem. However, the layout make 4-byte padding before `vals` on 64-bit CPU, so I would like to avoid it.

> If you are concerned about padding and want to keep this PR layout, you should be fine with a structure that defines only `size` and `capa_1`.

I agree with your example. Thank you!